util/codec/bytes.go

```diff
@@ -38,95 +21,93 @@ import (
 	"github.com/juju/errors"
 )
 
-// EncodeBytes encodes the slice value using an escape based encoding,
-// with the following rule:
-//  \x00 -> \x00\xFF
-//  \xFF -> \xFF\x00 if the first byte is \xFF
-// EncodeBytes will append \x00\x01 at the end of the encoded value to
-// indicate the termination.
+const (
+	encGroupSize     = 8
+	avgEncGroupCount = 20
+	encMarker        = byte(0xFF)
+	encPad           = byte(0x0)
+)
+
 // EncodeBytes guarantees the encoded value is in ascending order for comparison,
-// The encoded value is >= SmallestNoneNilValue and < InfiniteValue.
+// encoding with the following rule:
+//  [group1][marker1]...[groupN][markerN]
+//  group is 8 bytes slice which is padding with 0.
+//  marker is `0xFF - padding 0 count`
+// For example:
+//   [] -> [1 0 0 0 0 0 0 0 0 247]
+//   [1, 2, 3] -> [1 1 2 3 0 0 0 0 0 250]
+//   [1, 2, 3, 0] -> [1 1 2 3 0 0 0 0 0 251]
+//   [1, 2, 3, 4, 5, 6, 7, 8] -> [1 1 2 3 4 5 6 7 8 255 0 0 0 0 0 0 0 0 247]
+// Refer: https://github.com/facebook/mysql-5.6/wiki/MyRocks-record-format#memcomparable-format
 func EncodeBytes(b []byte, data []byte) []byte {
-	// Allocate more space to avoid unnecessary slice growing
-	bs := reallocBytes(b, len(data)+20)
-	if len(data) > 0 && data[0] == 0xFF {
-		// we must escape 0xFF here to guarantee encoded value < InfiniteValue \xFF\xFF.
-		bs = append(bs, 0xFF, 0x00)
-		data = data[1:]
-	}
-
-	for {
-		// find 0x00 and escape it.
-		i := bytes.IndexByte(data, 0x00)
-		if i == -1 {
-			break
+	// Allocate more space to avoid unnecessary slice growing.
+	// Assume that the byte slice size is about 20*8 = 160 bytes.
+	dl := len(data)
+	bs := reallocBytes(b, dl+avgEncGroupCount)
```

Comment:
now the slice size is not dl + 20, but for (len(data) / 8 + 1) * 9
see https://github.com/facebook/mysql-5.6/blob/webscalesql-5.6.24.97/storage/rocksdb/rdb_datadic.cc#L1255